src/os/bluestore/BlueStore.cc

```diff
@@ -5564,7 +5564,8 @@ void BlueStore::_do_write_small(
     uint64_t tail_read =
       ROUND_UP_TO(b_off + b_len, chunk_size) - (b_off + b_len);
     if ((head_read || tail_read) &&
-	b->get_ondisk_length() >= b_off + b_len + head_read + tail_read) {
+	(b->get_ondisk_length() >= b_off + b_len + tail_read) &&
+       b->is_allocated(b_off - head_read, b_len + head_read + tail_read)) {
```

Comment:
I don't think we want the is_allocated() check here unless we have a separate path to fill in zeros.  The read call will give us back zeros if it's not allocated so I don't think it's worth getting tricky here... let's keep it simple and just change the get_ondisk_legnth comparison bug!